### ISSUE TITLE
[feature request] Addition of custom loss capability to custom policies

### ISSUE THREAD

**Issue by karim174** [hill-a/stable-baselines#463]
i want to add a supervised loss to a custom model, as an auxiliary loss for my CNN i hope you add this as capability in future versions it's very useful as for addition of imitation learning losses on the during training like in [https://arxiv.org/abs/1804.02717](https://arxiv.org/abs/1804.02717) and many other RL algorithms.

**Comment by Miffyli** [comment]
I concur adding additional losses would be an useful option for the reasons karim174 said above (auxiliary losses, regularizers, etc), and in current form the code requires under-the-hood modifications to do this.

Perhaps a callback function in `setup_model` that provides locals and globals for users to play around with (much like the callback in `learn`)? Not the most elegant solution, but this would also allow other types of under-the-hood tinkering for model setup without actually changing the stable-baselines code.

**Comment by passant19** [comment]
i believe that would be a useful addition as for the mentioned reasons where it's needed for applications such as learning for improved dynamics and in vision task loss onto agent's CNN to encourage learning of useful features as in my application.

**Comment by araffin** [comment]
Hello,
some remarks about that feature:
Currently, because of tensorflow, all the losses are defined in `setup_model`, so to add new losses, the best and easiest way is to create a class that derives from a model (see https://github.com/araffin/learning-to-drive-in-5-minutes/blob/master/algos/custom_sac.py for an example)
@miffyli even though you provide a callback to the user, that does not solve the problem of data collecting/preprocessing, so i would not include that by default in stable baselines

**Comment by Miffyli** [comment]
Good point. There are some losses that could be included in existing `setup_model` functions without more tinkering, e.g. encoding/decoding the (crude) version of the image to have that autoencoder-style loss, or predicting immediate rewards as done in [UNREAL](https://arxiv.org/abs/1611.05397). 

But I agree: Even these are not universal across methods and I imagine each of them would require tinkering with the loss weights to get it right, so the best bet would be to create custom models as you suggested.
